tests/test_train.py

```diff
@@ -358,6 +363,258 @@ def test_train_nlu_autoconfig(
+        stories_yaml["stories"][1]["steps"].append({"user": "Yes"})
+
+        new_stories_file = tempfile.NamedTemporaryFile(mode="w", suffix=".yml").name
+
+        rasa.shared.utils.io.write_yaml(stories_yaml, new_stories_file)
+
+        mocked_nlu_training = mock_nlu_training(monkeypatch)
+        mocked_core_training = mock_core_training(monkeypatch)
+
+        new_model_path = train(
+            default_domain_path,
+            default_stack_config,
+            [new_stories_file, default_nlu_data],
+            output=new_model_path_in_same_dir(trained_e2e_model),
+        ).model
+        os.remove(new_model_path)
+
+        mocked_core_training.assert_called_once()
+        mocked_nlu_training.assert_not_called()
+
+    def test_nlu_and_core_trained_if_no_nlu_data_but_e2e_stories(
+        self,
+        monkeypatch: MonkeyPatch,
+        default_domain_path,
+        default_stack_config,
+        default_e2e_stories_file,
+        tmp_path,
+    ):
+
+        mocked_nlu_training = mock_nlu_training(monkeypatch)
+        mocked_core_training = mock_core_training(monkeypatch)
+
+        output = self.make_tmp_model_dir(tmp_path)
+        train(
+            default_domain_path,
+            default_stack_config,
+            [default_e2e_stories_file],
+            output=output,
+        )
+
+        mocked_core_training.assert_called_once()
+        mocked_nlu_training.assert_called_once()
+
+    def make_tmp_model_dir(self, tmp_path):
+        (tmp_path / "models").mkdir()
+        output = str(tmp_path / "models")
+        return output
+
+    def test_new_nlu_data_retrains_core_if_there_are_e2e_stories(
+        self,
+        monkeypatch: MonkeyPatch,
+        trained_e2e_model: Text,
+        default_domain_path,
+        default_stack_config,
+        default_e2e_stories_file,
+        default_nlu_data,
+    ):
+        nlu_yaml = rasa.shared.utils.io.read_yaml_file(default_nlu_data)
+        nlu_yaml["nlu"][0]["examples"] += "- surprise!\n"
+
+        new_nlu_file = tempfile.NamedTemporaryFile(mode="w", suffix=".yml").name
+
+        rasa.shared.utils.io.write_yaml(nlu_yaml, new_nlu_file)
+
+        mocked_nlu_training = mock_nlu_training(monkeypatch)
+        mocked_core_training = mock_core_training(monkeypatch)
+
+        new_model_path = train(
+            default_domain_path,
+            default_stack_config,
+            [default_e2e_stories_file, new_nlu_file],
+            output=new_model_path_in_same_dir(trained_e2e_model),
+        ).model
+        os.remove(new_model_path)
+
+        mocked_core_training.assert_called_once()
+        mocked_nlu_training.assert_called_once()
+
+    def test_new_nlu_data_does_not_retrain_core_if_there_are_no_e2e_stories(
+        self,
+        monkeypatch: MonkeyPatch,
+        trained_simple_rasa_model: Text,
+        default_domain_path,
+        default_stack_config,
+        simple_stories_file,
+        default_nlu_data,
```

Comment:
types

Reply:
@joejuzl could you please take this over